### DISCUSSION TITLE
Annotating method type for setImperativeHandle when using chakra-ui.forwardRef?

### DISCUSSION THREAD

**Discussion by rabelais88** [chakra-ui/chakra-ui#5160] [Q&A]
in React Hook, a method of component is exposed by `setImperativeHandle()`. Using `React.forwardRef` allows the type annotation of exposed method. However, it seems `React.forwardRef` and `@chakra-ui/system.forwardRef` has different generics.
```typescript
// React.forwardRef
function forwardRef<T, P = {}>(render: ForwardRefRenderFunction<T, P>): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<T>>;
// @chakra-ui/system.forwardRef
function forwardRef<Props extends object, Component extends As> { /* ... */ };
```
type annotation examples from Google search results are limited to `React.forwardRef`. 
- [sample example from Google search: stackoverflow](https://stackoverflow.com/questions/65300987/use-useimperativehandle-in-function-component-with-typescript-it-will-tips-some)

I've tried implementing the same stuff with chakra's `forwardRef` and it fails.
```typescript
import { forwardRef } from '@chakra-ui/system';
import { useImperativeHandle } from 'react';
import { Box } from '@chakra-ui/layout';
interface MyComponentProps { /* ... */ };
interface MyComponentHandle { doSomething: () => void };
const MyComponent = forwardRef<MyComponentProps, MyComponentHandle>(props, ref) => {
  useImperativeHandle(ref, () => {
    doSomething: function() { /* ... */ }
  })
  return <Box ref={ref} />
}
```

the `forwardRef` yields this type error:
```
Type 'MyComponentHandle' does not satisfy the constraint 'As<any>'.
  Type 'MyComponentHandle' is not assignable to type 'FunctionComponent<any>'.
```

accessing exposed method yields the error below:
```
Property 'doSomething' does not exist on type 'never'.ts(2339)
```

How can I annotate method type for `setImperativeHandle` when using chakra-ui.forwardRef?

**Comment by anubra266** [comment]
@rabelais88 create another internal ref for the box then merge them with useMergeRefs

Then use the internal ref in your imperative handle
```js
const internalRef = React.useRef()

const boxRef = useMergeRefs(internalRef, ref)

useImperativeHandle(internalRef, () => {
    doSomething: function() { /* ... */ }
})

return <Box ref={boxRef} />
```
